File: patterns/01_buidler/ex1_nutrifacts.java
package main;

// Step 1: Create a NutritionFacts class with members
public class NutritionFacts {
    private final int servingSize;
    private final int servings;
    private final int calories;
    private final int sodium;
    private final int carbohydrate;

    public class Builder {
        // Step 2: Require paramter
        // ===> Nghĩa là user cần phải input parameters.
        private final servingSize; 
        private final servings;

        // Step 3: Optional parameters, and initialize to default value.
        // ===> Nghĩa là user có quyền input hoặc không input các input parameters
        private int sodium;
        private int carbohydrate;

        // Step 4: Contructor
        // ===> Nghĩa là nếu cần object NutritioFacts, user phải sử dụng Constructor Builder này
        public Builder(int servingSize, int servings) {
            this.servingSize = servingSize;
            this.servings = servings;
        }

        // Step 5: Create methods with data type is Buidler class
        // ===> Nghĩa là user có thể sử dụng methods này để set parameters
        public Builder calories(int calories) {
            this.calories = calories;
            return this
        }

        // Step 5: 
        public Builder sodium(int sodium) {
            this.sodium = sodium;
            return this
        }

        // Step 5:
        public Builder carbohydrate(int carbohydrate) {
            this.carbohydrate = carbohydrate;
            return this
        }

        // Step 6: Create method with NutritionFacts class, with a input is a builder object
        public NutritionFacts build() {
            return new NutritionFacts(this);
        }
    }

    // Step 7: At main class, receive all input parameteres from Builder(0)
    private NutritionFacts(Buidler builder) {
        servingSize = builder.servingSize;
        servings = builder.servings;
        calories = builder.calories;
        sodium = builder.sodium;
        carbohydrate = builder.carbohydrate;
    }

    // Step 8: 
}

File: patterns/01_buidler/ex1_nutrifacts.go
package main

import "fmt"

// Step 1: Create Concrete class
// ===> Meaning that: from requirement, we must list all member into a structure.
type nutritionFacts struct {
	serveringSize int
	serverings    int
	calories      int
	sodium        int
	carbohydrate  int
}

// Step 2: Create methods
// ===> If you want to access to the above structure,
//		we can define all methods using interface as below
type NutritionFacts interface {
}

// Step 3: Create builder for NutrionFacts
// ===> Define t?t c? c?c methods cho vi?c kh?i t?o m?t object.
// 		Ph??ng ph?p n?y g?i l? Builder Pattern
type NutritionFactsBuilder interface {
	Serving(servingSize int, servings int) NutritionFactsBuilder
	Calories(calories int) NutritionFactsBuilder
	Sodium(sodium int) NutritionFactsBuilder
	Carbohydrate(carbohydrate int) NutritionFactsBuilder
	Build() NutritionFacts
}

// Step 4: Implement methods for builder
//		===> Kinh nghi?m cho th?y, ph?i implement h?t t?t c? c?c interface th? m?i kh?ng
//			 b? b?o l?i v? vi?c ch?a implement h?t c?c methods trong interface
func (nutri *nutritionFacts) Serving(servingSize int, servings int) NutritionFactsBuilder {
	nutri.serveringSize = servingSize
	nutri.serverings = servings
	return nutri
}

// Step 4:
func (nutri *nutritionFacts) Calories(calories int) NutritionFactsBuilder {
	nutri.calories = calories
	return nutri
}

// Step 4:
func (nutri *nutritionFacts) Sodium(sodium int) NutritionFactsBuilder {
	nutri.sodium = sodium
	return nutri
}

// Step 4:
func (nutri *nutritionFacts) Carbohydrate(carbohydrate int) NutritionFactsBuilder {
	nutri.carbohydrate = carbohydrate
	return nutri
}

// Step 5: Implement method Build() ?? set up h?t c?c paramter
// ===> Meaning that: Build() methods must return a reference pointer to object NutritionFacts
func (nutri *nutritionFacts) Build() NutritionFacts {
	return nutri
}

// Step 6: Must create a initial object from nutritionFacts
func NewNutritionFacts() NutritionFactsBuilder {
	// Important: This will return a variable from nutritionFacts
	return &nutritionFacts{}
}

func main() {
	// Case 1: Full parameter
	//		We must use NewNutritionFacts() to return a variable,
	//		then we can set parameter from input
	nutri1 := NewNutritionFacts().Serving(270, 80).Calories(100).Sodium(35).Carbohydrate(27).Build()
	fmt.Print(nutri1)
	fmt.Printf("\nEnd Case 1\n\n")

	// Case 2:
	nutri2 := NewNutritionFacts().Build()
	fmt.Print(nutri2)
	fmt.Printf("\nEnd Case 2\n\n")

	// Case 3:
	nutri3 := NewNutritionFacts().Serving(270, 80).Build()
	fmt.Print(nutri3)
	fmt.Printf("\nEnd Case 3\n\n")

}
